coq-8.17 branch no longer builds on ocaml 4.11.2

```
CAMLOPT -c -for-pack Template_coq src/quoter.ml
File "src/quoter.ml", line 345, characters 24-43:
345 |         let acc, arr' = Array.fold_left_map (fun acc t -> let t', acc = quote_term acc env sigma t in acc, t') acc ar in
                              ^^^^^^^^^^^^^^^^^^^
Error: Unbound value Array.fold_left_map
Command exited with non-zero status 2
src/quoter.cmx (real: 0.07, user: 0.06, sys: 0.00, mem: 26688 ko)
make[2]: *** [Makefile.template:743: src/quoter.cmx] Error 2
CAMLOPT -c -for-pack Template_coq src/denoter.ml
File "src/denoter.ml", line 166, characters 25-44:
166 |           let evm, arr = Array.fold_left_map (fun evm a -> aux env evm a) evm arr in
                               ^^^^^^^^^^^^^^^^^^^
Error: Unbound value Array.fold_left_map
Command exited with non-zero status 2
src/denoter.cmx (real: 0.06, user: 0.03, sys: 0.02, mem: 24656 ko)
make[2]: *** [Makefile.template:743: src/denoter.cmx] Error 2
```

Presumably this is due to https://github.com/MetaCoq/metacoq/pull/998 cc @mattam82 
Is there an easy way to restore compatibility so I don't have to recompile my whole 8.17 switch?

--- a/template-coq/src/denoter.ml
+++ b/template-coq/src/denoter.ml
@@ -163,7 +163,7 @@ struct
       | ACoq_tFloat x -> evm, Constr.mkFloat (D.unquote_float64 x)
       | ACoq_tArray (u, arr, def, ty) ->
           let evm, u = D.unquote_universe_level evm u in
-          let evm, arr = Array.fold_left_map (fun evm a -> aux env evm a) evm arr in
+          let evm, arr = ArrayCompat.fold_left_map (fun evm a -> aux env evm a) evm arr in
           let evm, def = aux env evm def in
           let evm, ty = aux env evm ty in
           evm, Constr.mkArray (Univ.Instance.of_array [|u|], arr, def, ty)

--- a/template-coq/src/quoter.ml
+++ b/template-coq/src/quoter.ml
@@ -342,7 +342,7 @@ struct
         let u = match Univ.Instance.to_array u with [| u |] -> u | _ -> assert false in
         let def', acc = quote_term acc env sigma def in
         let ty', acc = quote_term acc env sigma ty in
-        let acc, arr' = Array.fold_left_map (fun acc t -> let t', acc = quote_term acc env sigma t in acc, t') acc ar in
+        let acc, arr' = ArrayCompat.fold_left_map (fun acc t -> let t', acc = quote_term acc env sigma t in acc, t') acc ar in
         Q.mkArray (Q.quote_univ_level u) arr' ~default:def' ~ty:ty', acc
       in
       aux acc env trm

--- a/template-coq/src/tm_util.ml
+++ b/template-coq/src/tm_util.ml
@@ -314,3 +314,20 @@ type ('term, 'nat, 'ident, 'name, 'quoted_sort, 'cast_kind, 'kername, 'inductive
   | ACoq_tFloat of 'float64
   | ACoq_tArray of 'universe_level * 'term array * 'term * 'term
 
+module ArrayCompat = struct
+  (** For compat with OCaml < 4.13, copied from https://github.com/ocaml/ocaml/blob/d5b5c34971ed4291c385852f820554ad57adba84/stdlib/array.ml#L195-L207 *)
+  let fold_left_map f acc input_array =
+    let open Array in
+    let len = length input_array in
+    if len = 0 then (acc, [||]) else begin
+      let acc, elt = f acc (unsafe_get input_array 0) in
+      let output_array = make len elt in
+      let acc = ref acc in
+      for i = 1 to len - 1 do
+        let acc', elt = f !acc (unsafe_get input_array i) in
+        acc := acc';
+        unsafe_set output_array i elt;
+      done;
+      !acc, output_array
+    end
+end

